File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []

        for i in range(len(nums)):
            # 如果現在算的 nums[i] 跟前一個 nums[i-1] 是依樣的數字就跳過，避免重複計算
            if (i>0 and nums[i] == nums[i-1]):
                continue

            l = i+1
            r = len(nums)-1

            while l<r:
                total = nums[i] + nums[l] + nums[r]
                if total > 0:
                    r -= 1
                elif total < 0:
                    l += 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    # 手動把 l 往右邊移一格，不然會無限迴圈
                    l += 1
                    # l 要移動到跟前面push進去的 nums[l] 不一樣的數字，不然會重複計算一樣的case
                    while (nums[l] == nums[l-1] and l<r):
                        l += 1
            
        return res


        

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {
        function backtracking(idx, comb, total) {
            if (total === target) {
                res.push([...comb]);
                return;
            }

            if (total > target || idx >= candidates.length) {
                return;
            }

            comb.push(candidates[idx]);
            backtracking(idx+1, comb, total+candidates[idx]);
            comb.pop();

            while (idx+1 < candidates.length && candidates[idx] === candidates[idx+1]) {
                idx++;
            }
            backtracking(idx+1, comb, total);
        };

        let res = [];
        candidates.sort((a,b)=>a-b);
        backtracking(0, [], 0);
        return res;
    }
}
